File: frontend/vibematch_app/lib/features/home/models/home_banner.dart
import 'package:flutter/material.dart';

class HomeBanner {
  final String id;
  final String title;
  final String? imageUrl;
  final String placement;
  final String target;
  final String? targetUrl;
  final String? description;
  final int sortOrder;
  final bool isActive;
  final IconData fallbackIcon;
  final List<Color> fallbackGradient;

  const HomeBanner({
    required this.id,
    required this.title,
    this.imageUrl,
    this.placement = 'event',
    this.target = 'event',
    this.targetUrl,
    this.description,
    this.sortOrder = 1,
    this.isActive = true,
    required this.fallbackIcon,
    required this.fallbackGradient,
  });

  factory HomeBanner.fromJson(Map<String, dynamic> json) {
    final placement = json['placement']?.toString() ?? 'event';
    final target = json['target']?.toString() ?? (placement == 'policy_rules' ? 'policy' : 'event');
    return HomeBanner(
      id: json['id']?.toString() ?? '',
      title: json['title']?.toString() ?? 'Banner',
      imageUrl: _cleanString(json['image_url'] ?? json['imageUrl']),
      placement: placement,
      target: target,
      targetUrl: _cleanString(json['target_url'] ?? json['targetUrl']),
      description: _cleanString(json['description']),
      sortOrder: int.tryParse(json['sort_order']?.toString() ?? '') ?? 1,
      isActive: json['is_active'] != false,
      fallbackIcon: placement == 'policy_rules' ? Icons.rule_rounded : Icons.celebration_rounded,
      fallbackGradient: placement == 'policy_rules'
          ? const [Color(0xFF251538), Color(0xFF4A2A63)]
          : const [Color(0xFFE84C72), Color(0xFF8C5CF6)],
    );
  }

  static String? _cleanString(Object? value) {
    final text = value?.toString().trim();
    if (text == null || text.isEmpty || text == 'null') return null;
    return text;
  }
}

File: backend/app/schemas/home_banner.py
from datetime import datetime

from pydantic import BaseModel, Field


class HomeBannerResponse(BaseModel):
    id: int
    placement: str
    title: str
    image_url: str
    target: str
    target_url: str | None = None
    description: str | None = None
    sort_order: int
    is_active: bool
    starts_at: datetime | None = None
    ends_at: datetime | None = None

    model_config = {"from_attributes": True}


class HomeBannerCreateRequest(BaseModel):
    placement: str = Field(..., min_length=1, max_length=40)
    title: str = Field(..., min_length=1, max_length=160)
    image_url: str = Field(..., min_length=1, max_length=1000)
    target: str = Field(default="event", max_length=40)
    target_url: str | None = Field(default=None, max_length=1000)
    description: str | None = None
    sort_order: int = Field(default=1, ge=1)
    is_active: bool = True
    starts_at: datetime | None = None
    ends_at: datetime | None = None
